### ISSUE TITLE
Why pass in component name / how to deal with collisions?

### ISSUE THREAD

**Issue by bluenote10** [abishekatp/stylers#24]
First of all, thanks for creating this library, this looks really promising!

I'm wondering why the `style!` macro requires to pass in a component name?

Initially I thought this is used as the CSS class name, and I was worried how to prevent collisions. Looking a bit under the hood shows that the class name is already random (or hash) based, so the risk of a collision is small (or impossible?). Apparently the name is only used for the file name in `target/stylers/css`. But why not use simply use the same unique name as for the CSS style itself? If a project grows in size, or when mixing third party component libraries, a collision could happen at some point.

It looks like currently collisions of the files names are not detected / handled, right? At least creating a collision on purpose seems to simply overwrite the file, i.e., the resulting styles are broken.

**Comment by abishekatp** [comment]
Hi @bluenote10, Thank you for opening this issue. As you said, the class name we are using are randomly generated. **As you said indeed it is going to be a problem when project grows in size.**. But I am not sure about importing external components. Since both style and style_sheet macro generate css files at the compile time itself, component libraries may not be able to use these two macros. But that's why we have style_str! and style_sheet_str!. these two macros don't have this constraint and returns the scoped style as string.
- File names are used to uniquely identify component. We need this only in the case of style! or style_sheet! macros. To understand the need of this file name we must understand how we store the css.
- Whenever a leptos component got compiled it will call the style macro which will process the css either inside component itself or from external css file and store it in a separate css file. 
- One way to do it is directly appending the output of style macro to the final css file whenever style macro is compiled or recompiled. This will work very fine(when we use cargo build or trunk build) and also solve our problem. 
- But problem will arise when we use trunk serve. In this case not all the components will be recompiled each time. only the components which got changed going to be recompiled. So if we are just appending to the output css file means that will grow in size each time component got re compiled. 
- To avoid this problem we created unique css files for each component and override that file each time component got recompiled. 

- why can't we use randomly generated strings as file names? this is becase these strings are random so we will not be able identify the previously generated file name when componnet got recompiled. 


Solution:
If we are just going to use `trunk build` then we can avoid this component name constraint. 

**Problem: we need some logic to identiy the generated css file names for each component's style! macro.
If there is some better way to do it, then it will be very helpfull **

**Comment by bluenote10** [comment]
Thanks for the detailed explanation! This is indeed an interesting problem.

Another challenge I discovered when looking at the current implementation is that it currently an O(N^2) solution w.r.t. the number of style macro invocation, because each macro will basically re-iterate over all previously generated files, right? But fortunately one would have to have quite a lot of components until this becomes a true performance issue.

Looking around for a solution / alternatives I came across [this question](https://stackoverflow.com/a/68416360/1804173), which basically recommends against having side effects at all in proc macros, partially for the reason that invocation semantics are fuzzy, but also because they may become sandboxes eventually, which would make it impossible entirely.

That's tough, but perhaps it is possible to solve with a `build.rs` script (trunk always runs the `build.rs` even in `trunk serve`). I'd been playing around with them a little bit, and something like the following could work, even though it is also not super nice.

The main issue with a `build.rs` script is that we have to perform an extra parse step of the Rust sources, and use a heuristic to identify css style macro calls. We'd apply a sort of duck typing: If it looks like a css style macro call, it probably is a css style macro call. The `build.rs` could thus identify all css style macro calls, and generate the final css output file. Since this runs over everything matching e.g.  `src/**/*.rs` via globbing, it doesn't depend on what actually gets recompiled, i.e., we could always fully re-generate the css output file. Since parsing Rust sources should be orders of magnitude faster than actual compilation, it should be okay-ish if it always runs over everything. It also avoids the O(N^2) scalability issue.

Regarding the implementation of the proc macro itself, we would have to make sure that the css class names are generated deterministically from the information available at the call site. This could be done e.g. by hashing the module/file name including the line number. The crucial thing is that the proc macro and the `build.rs` script have to use exactly the same logic to compute the class name, so that they match eventually.

As a first proof-of-concept, I was experimenting with this `build.rs` (this is taken from another experiment of mine, so currently it only checks for top-level macro calls called `css!` and doesn't do any actual css parsing, but just to demonstrate the basic setup):

```rust
use glob::glob;
use std::{env::current_dir, fs};
use syn::parse::{Parse, ParseStream, Result};
use syn::Item;
use syn::{Ident, LitStr, Token};

// Getting debug output from a build.rs is not easy. Using a warning as a work-around,
// but that only allows single line output.
// https://github.com/rust-lang/cargo/issues/985#issuecomment-1071667472
// For full output use:
// less ./target/debug/build/*/output
macro_rules! p {
    ($($tokens: tt)*) => {
        println!("cargo:warning={}", format!($($tokens)*))
    }
}

fn main() {
    p!("Running build script...");
    p!("Working dir: {:?}", current_dir().unwrap());

    let pattern = format!("{}/src/**/*.rs", current_dir().unwrap().to_str().unwrap());
    for file in glob(&pattern).unwrap() {
        let file = file.unwrap();
        p!("{:?}", file);

        let content = fs::read_to_string(file).expect("Failed to read file");
        p!("{}", content);

        let ast = syn::parse_file(&content).unwrap();
        // p!("{:#?}", ast);

        for item in ast.items {
            if let Item::Macro(item_macro) = item {
                let path = &item_macro.mac.path;
                if path.leading_colon.is_none()
                    && path.segments.len() == 1
                    && path.segments[0].ident == "css"
                {
                    p!("Found css macro: {:#?}", item_macro);
                    //item_macro.mac.tokens.
                    let body = item_macro.mac.parse_body::<CssMacroCall>();
                    p!("Body: {:?}", body);
                }
            }
        }

        // let re = regex::Regex::new(r"css!\((.*)\)").unwrap();
        // let cap = re.captures(&content);
        // p!("{:#?}", cap);
    }
}

#[derive(Debug)]
struct CssMacroCall {
    css_var_name: String,
    css_content: String,
}

impl Parse for CssMacroCall {
    fn parse(input: ParseStream) -> Result<Self> {
        let css_var_name: Ident = input.parse()?;
        input.parse::<Token![,]>()?;
        let css_content: LitStr = input.parse()?;

        Ok(CssMacroCall {
            css_var_name: css_var_name.to_string(),
            css_content: css_content.value(),
        })
    }
}
```

What do you think about such an approach?

**Comment by abishekatp** [comment]
I think you have done deep dive into this problem and come up with this great idea. This approach seams very promising. Also as you have said reiterating all the file and rewriting them into single file for each invocation of macro is not a scalable idea. So this approach also solves that problem. 
- I have to checkout whether we can generate same unique class names in the procedural macro and also in procedure in build.rs. If that works means this will be great. I will try to implement it today and post the update here.

**Comment by BlackAsLight** [comment]
Any update on fixing this bug?

**Comment by abishekatp** [comment]
> Any update on fixing this bug?

Sorry I have been busy on some other things lately. May in a week or two I can come up with some basic version of this idea. Then we can proceed from there.

**Comment by abishekatp** [comment]
I have some doubts on this approach. As far as I can understand build.rs file is added to our own crate and will be run before that particular crate is built. Then how can the use that to export the functionality of stylers crate. 
Would it be good idea to create command line tool which can parse all the rust files and generate the scoped css?

**Comment by bluenote10** [comment]
> As far as I can understand build.rs file is added to our own crate and will be run before that particular crate is built.

I think a library can have their own `build.rs` as well and it will also get executed. So I'd assume both would be possible, either provide the `build.rs` via the stylers crate, or require users to integrate stylers by creating their own `build.rs`. For an inspiration of the latter look at e.g. the [cc-rs and its usage instructions](https://github.com/rust-lang/cc-rs#using-cc-rs). In this case the goal for the stylers crate would be to offer a convenient API allowing to write a similar one-liner `build.rs` on user site.

> The other idea I can think of:- Would it be good idea to create command line tool which can parse all the rust files and generate the scoped css?

Yes that could make sense. However from a user perspective it would become a bit inconvenient having to re-run the CSS generation manually. So sooner or later the user would want to create a `build.rs` anyway that then triggers the CLI tool to automate the CSS generation whenever their sources change. To make things more convenient it would make sense to directly offer an API that can be called from the `build.rs` to avoid having to go through a CLI. Without having put any real thoughts into it, perhaps something like:


```rs
// build.rs on user site

fn main() {
    stylers::CSSGenerator::new()
        .add_src_pattern("./src/**/*.rs")
        .generate("my_styles.css");
}
```

Having a CLI in addition the a `build.rs` interface may not hurt though, because there may also be use cases for manual invocation (even if it just simplifies debugging or testing the generator itself)?

--------

If the `build.rs` is rather provided by stylers itself, the main question becomes how to pass the e.g. the search path to it. I haven't looked into that yet, but perhaps it can either be done via environment variables, or via a configuration file that is discovered by convention (in the cargo/project root directory).

----------

There was also [a recent thread in the Rust forum](https://users.rust-lang.org/t/macro-with-side-effects/90548) on how to solve this problem without relying on side effects in proc macros, and the hints also go in the direction of integrating into `build.rs`. There doesn't seem to be a great solution out there, but at least it may serve as inspiration.

**Comment by abishekatp** [comment]
Thank you for the clear explanation, This makes more sense to me now. Once I have some working code I will soon update here soon.

**Comment by abishekatp** [comment]
Hi, it's been some time. So I have been trying to implement the approach we have been discussing here. So I have pushed the code here in [stylers_build](https://github.com/abishekatp/stylers/tree/stylers_build) branch. I am facing some problems. I hope we can get some solution here.

### As of now solved problems:
- First I extracted the core logic of stylers to separate stylers_core crate so we can use them independently. We have two kind of API's one that will accept CSS as a rust token stream namely from_ts(), another one that will accept CSS as string from_str().
- In stylers_build crate I am creating one exported function called build(). This will read all the files in the src directory and extract each style_build!() macro content from each component.

### Problems:
- Now to process this CSS content the simple way would be passing it as a string to from_str() function. But when we convert CSS token stream into a string, it just adds one space between each token instead of retaining the actual space information.
- Second approach would be passing token stream directly from_ts(). But complication here is that from_ts() internally uses span().unwrap() to get the space information from token stream. That means from _ts should be used only inside procedural macro. Then If we try to use the macro directly to process the content then How can we pass this token stream to the macro?

**If we can solve any one of these problems then we can adapt to this build approach**. I might be wrongly assuming something here. So please let me know If I am wrong.

**Comment by abishekatp** [comment]
Hi @bluenote10 it's been long time. Finally I think I have figured out some way to achieving what we have discussed in this issue. Most probably in this week I will release the beta version of this implementation.

**Comment by BlackAsLight** [comment]
This issue can probably be closed now?
